### ISSUE TITLE
Model Builder can't handle large CSV files

### ISSUE THREAD

**Issue by andrasfuchs** [dotnet/machinelearning-modelbuilder#2195]
**System Information (please complete the following information):**
 - Model Builder Version (available in Manage Extensions dialog): 16.13.9.2231505
 - Visual Studio Version: 17.2.5 Enterprise

**Describe the bug**
- On which step of the process did you run into an issue: 
Right after I tried to load my large CSV file as data source.
- Clear description of the problem:
I usually work with large CSV files with many columns. This particular file has 150'000 feature columns and its size is 5.8 GB. I made a greatly reduced version of it which has only 100 data rows, that is only 200 MBs.

You can download them from my Google Drive: 
- [Zipped 5.8 GB CSV with generated .mbconfig (~1.3 GB)](https://drive.google.com/file/d/1z2xup6Mlu0IwwoKlXliLGKdi7Sn46C7l/view?usp=sharing)
- [Zipped 200 MB CSV with generated .mbconfig (~45 MB)](https://drive.google.com/file/d/1yzbu2jjNmvSnVSSgZDqnZ5BDUZDLTmlG/view?usp=sharing)

Please note, that in order to do some test with them, you will need to change the DataSource:FilePath value in the .mbconfig file, because they are specific to my directory structure.

When you try to open those files in Visual Studio, Model Builder uses ~20% CPU constantly, seemingly endlessly. **I waited 95 minutes** using the smaller, 200 MB file, but the Train tab didn't get enabled, so I suppose the file was not loaded until that point.
(If I don't open the .mbconfig file/Model Builder tab, the CPU usage is <1%.)

**To Reproduce**
1. Go to [my shared Google Drive folder](https://drive.google.com/drive/folders/1Kev-3jBDqkrtJOH7cMN0_-6AfFFfhNm1) and download the two zip files named `BBD_20220711_TrainingData_MLP06_1p0Hz-150000Hz_IsActivity_Meditation_*.zip`
2. Unzip them
3. Change the FilePath value in the .mbconfig files
4. Create an empty C# console application in Visual Studio
5. Copy the .mbconfig files into the folder of that console application
6. Include them and try to open them
7. The Model Builder starts to open the large CSV files, but it gets stuck 

**Expected behavior**
Opening huge files should be quick enough to start training in a reasonable time. Theoretically Model Builder should only read the first few lines of these big files to detect their structure, the data rows are not needed until we start training.

**Screenshots**
![image](https://user-images.githubusercontent.com/910321/178314239-435c5875-2cd1-49d8-a065-f95a5256f930.png)

![image](https://user-images.githubusercontent.com/910321/178314328-da538efe-f229-4ffc-ab6d-b883a12209b2.png)

**Additional context**
There is another, related issue I created in January that is now closed, but not completely solved: #1986

**Comment by andrasfuchs** [comment]
@LittleLittleCloud @luisquintanilla Sorry for mentioning you guys directly in this issue, but I think it highlights an important bottleneck ML.NET and/or Model Builder has when it comes to big data files. 

Could you test it too and confirm that it's a valid problem before the issue becomes stale/closed (again)?

I can give you other, similar files to do tests with, or I can even screenshare my typical training session live if you preferred that.

Thank you!

**Comment by LittleLittleCloud** [comment]
@beccamc, will your new PR on improving parsing large CSV in data page helps improve this scenario.

**Comment by luisquintanilla** [comment]
Hi @andrasfuchs 

Thanks for your patience. Did you run into this issue? I tried with the full file and it can't load it (I assume because of missing data or parsin issues). 

![image](https://user-images.githubusercontent.com/46974588/181612681-e056c5fb-ac5c-48f1-9d10-1360cc54771c.png)

**Comment by andrasfuchs** [comment]
@luisquintanilla I have just tested it again, and I'm still in the loading state after 15 minutes. The `Train` tab is inactive and my CPU is constantly ~20%. For how long did you wait after opening the file until you got that error?

**Comment by luisquintanilla** [comment]
> @luisquintanilla I have just tested it again with the bigger file, and I'm still in the loading state after 110 minutes. The `Train` tab is inactive and my CPU is constantly ~20%. For how long did you wait after opening the file until you got that error?

@andrasfuchs the error was immediately thrown. The file never loaded in Model Builder.

**Comment by andrasfuchs** [comment]
@luisquintanilla Strange, it works fine on my machine. Your error message seemed familiar, and I realized that I had the same problem before, back in January. 
It is a closed issue now, so it supposed to be solved: https://github.com/dotnet/machinelearning-modelbuilder/issues/1986

**Comment by luisquintanilla** [comment]
Thanks for referencing that issue @andrasfuchs. I've updated my version of Model Builder to  16.13.9.2237801 but still get the same issue. I'll try an internal Build to see if the issue persists.

**Comment by andrasfuchs** [comment]
@luisquintanilla Did you have the chance to try this with an internal build?

**Comment by luisquintanilla** [comment]
@andrasfuchs I did and I'm getting the same error as before. We'll investigate some more on our end but I agree with your initial point that for large files although the training should take a significant amount of time, data preview should be quick.

**Comment by beccamc** [comment]
@luisquintanilla I think the problem you're seeing is a timeout from PROSE. IIRC it just returns a null object when it times out, which we can't distinguish from a null result. I'll double check the code because if its a timeout we should be showing a different error

**Comment by luisquintanilla** [comment]
Hi @andrasfuchs 

For the issue you reported of MB taking a long time to load a file with 150k columns I believe we narrowed it down to the checks we perform using PROSE to detect and set column types (@beccamc to confirm). Other than working with the PROSE team to improve these checks, the other alternative is to disable them which would mean you'd have to manually set the data types for each of the 150k columns. Since 150k columns is significantly more than the usual number of columns we typically see in Model Builder, there is no plan to prioritize optimizing for this edge case at this time. As we hear more feedback on wide datasets like yours, we can revisit this topic.

In the meantime, what we can do is provide you with a notebook sample containing the code that can make it easier for you to load your dataset using the API. Would that help unblock you?

**Comment by andrasfuchs** [comment]
@luisquintanilla It's understandable.

Thank you, yes, a notebook sample would be great, if you don't mind creating one!

**Comment by luisquintanilla** [comment]
Hi @andrasfuchs 

Here you go. 

https://gist.github.com/luisquintanilla/a7fb1eb3ce00685a29f6fcd25abe1a42

Some things you might have to change are:

1. Update `dataPath` with the location of your data
2. Remove shuffling and sampling. I did those to quickly validate the pipeline I set up worked on a small subset (5k rows) of the full dataset.
3. Update the number of seconds to train for. Since you're using the full dataset, I'd recommend at least 1-2 hours (3600-7200s). The nice thing is that since you're not running this inside VS, you can leave it running in the background. 

Closing this issue for now. Feel free to reach out if you have questions or issues.

**Comment by andrasfuchs** [comment]
First of all, thank you @luisquintanilla for the notebook, it worked like a charm!

Good news is that I was also able to load the big CSV file into Model Builder after the #2132 fix! 
Here are my quick notes about my 5-hour-long training process, you might or might not find it useful :)

```
* Model Builder loaded the 33 MB .mbconfig in 19 seconds
* Limited to FASTTREE only
* Trainer process used 4.2GB RAM and the CPU remained at ~11% during training
* The generated C# class is 44 MB big. -> Visual Studio hangs when I tried to open it and scroll in that file, even with https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.editorbrowsablestate?view=net-6.0 attribute that theoretically would let IntelliSense skip the class.


Evaluate Metric: RSquared
Available Trainers: FASTTREE
Training time in second: 18000

|      Trainer                             RSquared   Duration       |
|--------------------------------------------------------------------|
|0     FastTreeRegression                  0,3767     2095,9120      |
|1     FastTreeTweedieRegression           0,8931     2218,2240      |
|2     FastTreeRegression                  0,2872     2268,8880      |
|3     FastTreeTweedieRegression           0,9585     2272,4400      |
|4     FastTreeRegression                  0,8163     2143,9420      |
|5     FastTreeTweedieRegression           -0,9020    2042,1170      |
|6     FastTreeRegression                  0,8821     2065,2580      |
|7     FastTreeTweedieRegression           0,6253     2135,6030      |
|--------------------------------------------------------------------|
|                          Experiment Results                        |
|--------------------------------------------------------------------|
|                               Summary                              |
|--------------------------------------------------------------------|
|ML Task: regression                                                 |
|Dataset: o:\Work\BBD.BodyMonitor\BBD_20220711_TrainingData_MLP06_1p0Hz-150000Hz_IsActivity_Meditation.csv|
|Label : IsActivity_Meditation                                       |
|Total experiment time : 18383,0000 Secs                             |
|Total number of models explored: 9                                  |
|--------------------------------------------------------------------|
|                        Top 5 models explored                       |
|--------------------------------------------------------------------|
|      Trainer                             RSquared   Duration       |
|--------------------------------------------------------------------|
|3     FastTreeTweedieRegression           0,9585     2272,4400      |
|1     FastTreeTweedieRegression           0,8931     2218,2240      |
|6     FastTreeRegression                  0,8821     2065,2580      |
|4     FastTreeRegression                  0,8163     2143,9420      |
|7     FastTreeTweedieRegression           0,6253     2135,6030      |
|--------------------------------------------------------------------|

```

**Comment by luisquintanilla** [comment]
@andrasfuchs great to hear it worked and as always, thanks for the detailed writeup. A few questions:

1. I see the scenario you used was value prediction (regression), but I thought based on the data you provided it was a classification problem (trying to predict the value of the last column to determine whether the activity is meditation). Did I misunderstand the problem?
2. When you say generated C# file, I assume you mean the *\*.training.cs* file. Is that right?

**Comment by andrasfuchs** [comment]
@luisquintanilla
1. Your assumption was right, but I prefer to use regression for now, because I'm switching from mental state detection to more numerical values (eg. medical values gained from blood samples). My understanding is that it's not wrong to use value prediction in case of meditation detection either, because there can be different levels to meditation, I'm just not able to detect them currently, that's why I use only 0 and 1 at the moment.
2. I can't remember if it was the _*.training.cs_ or the _*.consumption.cs_, but I think both grew quite big because of the many named properties that represented the columns in the CSV. I think it will be not an issue as soon as I load the model zip manually and start using your recommended way of handling my inputs and outputs that I saw [in your notebook](https://gist.github.com/luisquintanilla/a7fb1eb3ce00685a29f6fcd25abe1a42?short_path=a341a04#file-largedataset-ipynb-L84).

**Comment by luisquintanilla** [comment]
Makes sense. Thanks for clarifying.

**Comment by andrasfuchs** [comment]
I had a productive week last week, so I thought I share a few of my notes again with you, I think it has some useful insights for the Model Builder team. The #2132 fix finally unblocked me, thank you again for that :) I used Model Builder v16.13.10 for my tests and I limited to algorithm to FastTreeRegression only.

So, I moved from console application to a locally run web API app that has a singleton service that manages the ML training. I modified my data preparation a little too, so now I have more control over the quantity and quality of the data in my CSV files. 
My goal was to find the good model quality / training time ratio for my use case. I used both Model Builder and AutoML, but I will include only Model Builder results here.

1. Balanced, randomized CSV data source with ~3800 rows
MLP06: 7.37 GB, 150k columns, 2062 s/model
MLP07: 0.39 GB,   8k columns,   30 s/model
MLP08: 0.73 GB,  15k columns,   42 s/model
MLP09: 1.47 GB,  30k columns,  **221 s/model**
MLP10: 1.23 GB,  25k columns,   74 s/model

2. Balanced, randomized CSV data source with 1000 rows
MLP09: 0.39 GB, 30k columns,  52 s/model
MLP10: 0.33 GB, 25k columns,  39 s/model
MLP11: 1.24 GB, **95k columns**, **583 s/model**

3. Balanced, randomized CSV data source trained for 30 minutes
MLP09:  200 rows, 0.08 GB, 30k columns,  65 s/model, max. RSquared 0.7962 as 16th, 27 models tested in 30 minutes
MLP09: 1000 rows, 0.39 GB, 30k columns,  72 s/model, max. RSquared 0.9373 as 5th, 22 models tested in 30 minutes
MLP09: 3500 rows, 1.37 GB, 30k columns,  91 s/model, max. RSquared 0.9329 as 14th, 20 models tested in 30 minutes
MLP10: 3500 rows, 1.14 GB, 25k columns,  83 s/model, max. RSquared 0.8036 as 8th, 21 models tested in 30 minutes

I'm sure the exact numbers are fairly system resource-dependent, but according to my measurements having less than 25k columns and smaller than ~1.2 GB of data significantly lowers the training time per trial. Having more columns also generate huge C# files in the end that can freeze Visual Studio.

I also noticed that both in Model Builder and AutoML the training times are very similar for the consecutive trials with the same data source, but every now and then there is one or two trials that run significantly longer. This might be a bug or memory issue, I will create a detailed report in the AutoML repo later.

I just wanted to share all this with you, so that you can compare these results with your own, and maybe adjust the default settings of Model Builder, or show warnings if the user exceeds the reasonable limits.
